File: dart/hello_ffi.dart
import 'dart:ffi';
import 'pact_ffi_bindings.dart';
import 'package:ffi/ffi.dart';
import 'dart:io' show Platform, Directory;

import 'package:path/path.dart' as path;

void main(List<String> args) async {
  // print('hello ffi');
  var libraryPath = path.join(Directory.current.path, 'libpact_ffi.so');
  if (Platform.isMacOS) {
    libraryPath = path.join(Directory.current.path, 'libpact_ffi.dylib');
  } else if (Platform.isWindows) {
    libraryPath = path.join(Directory.current.path, 'pact_ffi.dll');
  }
  final lib = PactFfiBindings(DynamicLibrary.open(libraryPath));

  final version = lib.pactffi_version().cast<Utf8>().toDartString();
  lib.pactffi_logger_attach_sink('stdout'.toNativeUtf8().cast<Char>(),5);
  lib.pactffi_logger_apply();
  lib.pactffi_log_message('pact-dart-ffi'.toNativeUtf8().cast<Char>(), 'INFO'.toNativeUtf8().cast<Char>(), "hello from ffi version: $version".toNativeUtf8().cast<Char>());
}

File: java/jna/src/ffi/example/jna/HelloFfi.java
package msk.example.jna;

import com.sun.jna.Library;
import com.sun.jna.Native;
import com.sun.jna.Pointer;
import com.sun.jna.Structure;
import com.sun.jna.Platform;

public class HelloFfi {

    public interface PactFfi extends Library {
        PactFfi INSTANCE = (PactFfi) Native.load((Platform.isWindows() ? "pact_ffi.dll" : Platform.isLinux() ? "libpact_ffi.so" : "libpact_ffi.dylib"),
                PactFfi.class);

        String pactffi_version();

        Void pactffi_logger_init();

        int pactffi_logger_attach_sink(String sinkSpecifier, int LevelFilter);

        Void pactffi_logger_apply();

        Void pactffi_log_message(String source, String logLevel, String message);
    }

    public static void main(String[] args) {
        var version = PactFfi.INSTANCE.pactffi_version();
        PactFfi.INSTANCE.pactffi_logger_init();
        PactFfi.INSTANCE.pactffi_logger_attach_sink("stdout",3);
        PactFfi.INSTANCE.pactffi_logger_apply();
        PactFfi.INSTANCE.pactffi_log_message("pact-java-jna","INFO","hello from ffi version: "+version.toString());
    }
}